src/kv/tx.h

```diff
@@ -241,37 +248,34 @@ namespace kv
         throw std::logic_error("Transaction aborted");
 
       // If no transactions made changes, return a zero length vector.
-      bool changes = false;
+      const bool any_changes =
+        std::any_of(view_list.begin(), view_list.end(), [](const auto& it) {
+          return it.second.view->has_changes();
+        });
 
-      for (auto it = view_list.begin(); it != view_list.end(); ++it)
-      {
-        if (it->second.view->has_changes())
-        {
-          changes = true;
-          break;
-        }
-      }
-
-      if (!changes)
+      if (!any_changes)
       {
         return {};
       }
+
       // Retrieve encryptor.
       auto map = view_list.begin()->second.map;
       auto e = map->get_store()->get_encryptor();
 
       KvStoreSerialiser replicated_serialiser(e, version);
-      // flags that indicate if we have actually written any data in the
-      // serializers
-      auto grouped_views = get_views_grouped_by_domain(view_list);
 
-      for (auto domain_it : grouped_views)
+      // Process in security domain order
+      for (auto domain : {SecurityDomain::PUBLIC, SecurityDomain::PRIVATE})
```

Comment:
Replaced `get_views_grouped_by_domain` with this loop, which I think is equivalent. We iterate multiple times (here, and for `has_changes` above), and this could be flattened into a single pass building an ordered `view_list` of serialise-worth views, but we don't need to.